obhod_lune/obhod.py

```diff
@@ -0,0 +1,533 @@
+        self.mu_line = mu_line
+
+        self.dx = start_dx
+        self.dy = start_dy
+        self.dz = 0.0
+
+    # Distance to Earth
+    def R(self):
+        return math.sqrt(math.pow(self.x + self.mu, 2)
+                         + math.pow(self.y, 2)
+                         + math.pow(self.z, 2))
+
+    # Distance to Moon
+    def r(self):
+        return math.sqrt(math.pow(self.x - self.mu_line, 2)
+                         + math.pow(self.y, 2)
+                         + math.pow(self.z, 2))
+
+    # Calculate the 2nd derivative values at current state
+    def get_2nd_derivatives(self, dx, dy) :
+        ddx = self.x \
+              + 2 * dy \
+              - (self.x + self.mu) * self.mu_line / math.pow(self.R(), 3) \
+              - (self.x - self.mu_line) * self.mu / math.pow(self.r(), 3)
+
+        ddy = self.y \
+              - 2 * dx \
+              - self.y * self.mu_line / math.pow(self.R(), 3) \
+              - self.y * self.mu / math.pow(self.r(), 3)
+
+        ddz = - self.mu_line / math.pow(self.R(), 3) * self.z \
+              - self.mu / math.pow(self.r(), 3) * self.z
+
+        return ddx, ddy, ddz
+
+    # Apply RK4, and update values accordingly
+    def make_step(self, h) :
+        ddx1, ddy1, ddz1 = self.get_2nd_derivatives(self.dx, self.dy)
+
+        dx1 = self.dx + h / 2 * ddx1
+        dy1 = self.dy + h / 2 * ddy1
+
+        ddx2, ddy2, ddz2 = self.get_2nd_derivatives(dx1, dy1)
+
+        dx2 = self.dx + h / 2 * ddx2
+        dy2 = self.dy + h / 2 * ddy2
+
+        ddx3, ddy3, ddz3 = self.get_2nd_derivatives(dx2, dy2)
+
+        dx3 = self.dx + h * ddx3
+        dy3 = self.dy + h * ddy3
+
+        ddx4, ddy4, ddz4 = self.get_2nd_derivatives(dx3, dy3)
+
+        ddx = ddx1 + 2 * ddx2 + 2 * ddx3 + ddx4
+        ddy = ddy1 + 2 * ddy2 + 2 * ddy3 + ddy4
+        ddz = ddz1 + 2 * ddz2 + 2 * ddz3 + ddz4
+
+        self.dx = self.dx + h / 6 * ddx
+        self.dy = self.dy + h / 6 * ddy
+        self.dz = self.dz + h / 6 * ddz
+
+        self.x = self.x + h * self.dx
+        self.y = self.y + h * self.dy
+        self.z = self.z + h * self.dz
+
+        return self.x, self.y, self.z
+
+
+# Use bisection to find optimal power for translunar return trajectory
+# This should return a trajectory with an 8 shape
+def power_bisection_translunar(h, mu, mu_line, start_x, return_x, low_dy=10.7, high_dy=10.8, max_bisections=50):
+    for i in range(max_bisections):
+        print('Bisection {}'.format(i + 1))
+        mid_dy = (low_dy + high_dy) / 2
+
+        satellite = Satellite(start_x=start_x, start_y=0.0, start_dx=0.0, start_dy=mid_dy, mu=mu, mu_line=mu_line)
+
+        too_low = False
+        too_high = False
+
+        i_passed = False
+        ii_passed = False
+        iii_passed = False
+        iv_passed = False
+        v_passed = False
+        vi_passed = False
+
+        iv_y_acceleration_switch = False
+
+        while True:
+            min_dist = min(satellite.R(), satellite.r())
```

Comment:
This loop has high complexity. Again, I would rather formulate the condition in terms of a mathematical equation and use generic bisection method to solve it. This way it is much easier to replace bisection with some other method and debug the code to see where things go wrong.